📦 NEW: Add support for script fields and runtime fields

michaelborn opened the pull request:
This allows developers to return a custom value calculated for each document hit:

```js
searchBuilder.setScriptFields( {
    "interestCost": {
        "script": {
            "lang": "painless",
            "source": "return doc['price'].size() != 0 ? doc['price'].value * (params.interestRate/100) : null ",
            "params": { "interestRate": 5.5 }
        }
    }
} );
searchBuilder.setSource( true );
...
var interest = searchBuilder.getHits().map( (document) => document.getScriptFields()["interestCost"] );
```

- [x] Includes test 🤖
- [x] Includes docs 📖

michaelborn committed:
📦 NEW: Add script field support for dynamic runtime fields

This allows developers to return a custom value calculated for each document hit.

michaelborn committed:
📖 DOC: Add docs for new script field support

michaelborn commented:
@jclausen I created a PR just to get some eyes on this. I'm a bit stuck on this issue where the `_source` field is not returned in the search result when any script field is set.

![Screenshot from 2023-03-23 23-59-18](https://user-images.githubusercontent.com/8106227/227421050-af568eac-a60a-419f-82ce-112c81ec88b4.png)

I can't find anything in the docs mentioning this, and I don't see the issue in the client app I grabbed this feature from. Any ideas?

jclausen commented on docs/Searching/Search.md:
```diff
@@ -121,6 +121,38 @@ searchBuilder.sort( "author.age", "DESC" );
+        }
+    }
+} );
+searchBuilder.setSource( true );
```
My suggestion would be just to be opinionated about setting the source param ( e.g. make it an argument to `setScriptFields` and default it to `true`.  Then, you can just set the param when you apply the script field.

jclausen replied:
Oh.  I see. You are only handling this an accessor.  I think an overload would make sense, in this case, to allow for the source argument.

jclausen commented on models/SearchBuilder.cfc:
```diff
@@ -50,6 +50,13 @@ component accessors="true" {
+	 * 
+	 * https://www.elastic.co/guide/en/elasticsearch/reference/current/search-fields.html#script-fields
+	 */
+    property name="scriptFields" type="struct";
```
In addition to this being an accessor, I would like to see an `addScriptField` method that allows you to append to the struct.

jclausen replied:
I don't see the add method here.

michaelborn replied:
https://github.com/coldbox-modules/cbelasticsearch/blob/90386f202ab228c444d3fa3b2b958971e83f664d/models/SearchBuilder.cfc#L1195-L1209

jclausen commented on docs/Searching/Search.md:
````diff
@@ -121,6 +121,38 @@ searchBuilder.sort( "author.age", "DESC" );
+var interest = searchBuilder.getHits().map( (document) => document.getScriptFields()["interestCost"] ); // 5.50
+```
+
+How *interesting*. 😜
````
No editorializing in the docs! ;-P

jclausen commented on models/Document.cfc:
```diff
@@ -46,6 +46,10 @@ component accessors="true" {
+	/**
+	 * Any evaluated script field results
+	 */
+	property name="scriptFields" type="struct";
```
If we are going to implement this for script fields, let's add handling for runtime fields at the same time ( https://www.elastic.co/guide/en/elasticsearch/reference/current/runtime-retrieving-fields.html ).

For both, I feel like we need `getMemento` to include those - or at least allow an argument to flatten them together in to a single memento.

jclausen replied:
You marked this as resolved, but I don't see the runtime fields support added anywhere.

michaelborn replied:
Elasticsearch is confusing in that script fields are `"script_fields"`, but runtime fields are just `"fields"`. :facepalm: But you're right, I missed adding a `addField()` appender method for runtime/a la carte document fields to return.

The runtime field support on the response side is pretty straightforward, and is retrieved the same as script fields.

From the docs:

---

In a similar fashion, runtime fields can also be fetched and will appear in the document's `fields` struct. This example retrieves the `"fuel_usage_in_mpg"` runtime field as well as the indexed `"make"` and `"model"` fields:

```js
var hits = searchBuilder.new( "itinerary" )
             .setFields( [ "fuel_usage_in_mpg", "make", "model" ] )
             .execute()
             .getHits();
for( hit in hits ){
    var result = hit.getFields();
    writeOutput( "This #result.make# #result.model# gets #fuel_mpg#/gallon" );
}
```

To access document `fields` as well as the `_source` properties, use`hit.getDocument( includeFields = true)`:

```js
var result = searchBuilder.execute();
for( hit in result.getHits() ){
    var document = document.getDocuments( includeFields = true );
    writeOutput( "This #document.make# #document.model# gets #fuel_mpg#/gallon" );
}
```

jclausen reviewed (changes requested):
Some thoughts.  I'd like to see us add support for requesting runtime fields and handling them in both the search builder and document objects.

michaelborn committed:
👌 IMPROVE: Update script field support per Jon's feedback

1. Use an `addScriptField()` method for appending, rather than overwriting, script fields.
2. Set the `_source` to true by default, to avoid missing `_source` key errors when reading results.

michaelborn committed:
Merge branch 'development' of github.com:coldbox-modules/cbelasticsearch into feat/add-script-field-support

michaelborn committed:
👌 IMPROVE: Allow separate _source_ and fields document structs

1. Allows creating a runtime mapping through the index builder via the `"mappings.runtime"` property.
2. Fixes bug where search result population bombs if `_source` does not exist. (Which can happen for a script field.)
3. Rename `memento` to `_source` in Document.
4. Support `includeFields = true` for acquiring a merged struct of _source and fields from the document. (False by default.)

michaelborn committed:
📖 DOC: Add docs for runtime fields

michaelborn committed:
🤖 TEST: Speed up HyperClient tests by 50 seconds

We don't (seem to) need the `sleep(500)` calls in between tests, so why have them there? The client tests go from 65s runtime to 15s with the sleep calls removed.

michaelborn commented:
I also added a commit to drop all the `sleep( 500 )` calls in between tests, which speeds up the `HyperClientTest` by about 50 seconds! All the tests pass on my build... I'm hoping this is also the case for the ACF builds and dual cluster tests. 🤞


_compare..._
![Screenshot from 2023-03-24 15-16-13](https://user-images.githubusercontent.com/8106227/227623937-5acbab1e-88fc-4dd4-bd36-7148d0804d83.png)

_to..._
![Screenshot from 2023-03-24 15-14-15](https://user-images.githubusercontent.com/8106227/227623942-b4130544-90ae-4099-a9bc-30aa8609d993.png)

michaelborn committed:
Merge branch 'development' of github.com:coldbox-modules/cbelasticsearch into feat/add-script-field-support

michaelborn committed:
👌 IMPROVE: Add convenience methods and better documentation for runtime fields

michaelborn commented:
* I added an `addField()` convenience method for appending a field name or config struct to the list of fields to return.
* I improved the documentation concerning runtime field retrieval
* I added a test for the `.addField()` notation
